### ISSUE TITLE
Extracting and de-duping nested models

### ISSUE THREAD

**Issue by eropple** [jayalfredprufrock/nestjs-typebox#50]
This was surprising behavior, and I'm still chewing over how to deal with it. Here's a stripped-down example of the models I'm working with:

```ts
export const UserPrivate = Type.Object({
  id: Type.String(),
  displayName: Type.String(),
  email: Type.String(),
  verified: Type.Boolean()
});
export type UserPrivate = Static<typeof UserPrivate>;

export const WhoamiDto = Type.Object({
  youAre: Type.Union([
    UserPrivate,
    Type.Null(),
  ]),
})
export type WhoamiDto = Static<typeof WhoamiDto>;
```

My assumption (and I think NestJS's stock class-transformer/class-validator soup would do this? but I'm not sure) was that each object would be pulled out so that `components.schemas` would have two separate fields, but...obviously that's not easily possible without more information. Could we come up with a way to standardize on providing that information? When I wrote [nestjs-openapi3](https://github.com/eropple/nestjs-openapi3) I wasn't really thinking about things like Typebox at all, but my [fastify-openapi3](https://github.com/eropple/fastify-openapi3), tackled this through schema tagging, e.g.:

```ts
export const SCHEMA_NAME_PROPERTY = "x-fastify-schemaName" as const;

/**
 * Decorates a `@sinclair/typebox` schema with a name for use in
 * `@eropple/fastify-openapi3`.
 *
 * If you decorate two separate types with the same `name`, this
 * will explode at speed. You have been warned. I will laugh.
 */
export function schemaType<T extends TSchema & CustomOptions>(
  name: string,
  type: T
): T & TaggedSchema {
  return { ...type, [SCHEMA_NAME_PROPERTY]: Symbol(pascalCase(name)) };
}
```

Any thoughts? If this library were to go that way, there's probably a good bit of reusable code in fastify-openapi3. :)

Thanks for your consideration!

**Comment by smiley-uriux** [comment]
I managed to miss this issue until now, but I'm open to exploring things with you, although I'm not 100% positive I understand your concern and how it relates to this library. My use case thus far has been mostly focused on validation, but the library does provide basic integration with the swagger/openapi NestJs plugin for spec generation (by simply passing off the generated JSON schema produced by Typebox to existing NestJs tooling). That said, any limitations in terms of how references and unions are handled are likely outside the scope of this library. Typebox does provide support for schema refs, and assuming the NestJs swagger plugin handles them correctly, it might be possible to achieve the schema reuse you are looking for that way. For my use cases, the expansion of model variants as separate schemas isn't problematic, and in many cases desirable, since I've found openapi client and documentation generators to do a better job with explicit/expanded schemas versus layers of references. I'm guessing validation performance would suffer as well.

I might be completely misunderstanding things here, so forgive me if so. If you'd like to provide more context I'm happy to continue the conversation and will certainly be more responsive as I'm still actively maintaining this project and have some big plans for a 3.0 release. I'll keep this issue open until I hear back from you. Thanks!
